hold ref to temp keychain on OSX to avoid premature cleanup

wfurt opened the pull request:
This is fix for issue originally reported on macOS 11.0 when we fail to add certificates to a store. 
The root cause if fact that when importing, we put certificate and key to temporary keyChain and that can be deleted before we use the bits. This used to work with 3.1 but it is failing now on Catalina+. 

The fix is to grab reference when temporary keychain is used to make sure keychain file stays around. 
Since the chain can now linger longer, all the changes in `CreateTemporaryKeychain` are to ensure we have better one time encryption password. Technically, that is not needed to avoid the issue.

As far as testing, this is little bit tricky. We already have tests that would expose the behavior.
However, the X509StoreMutableTests are guarded by PermissionsAllowStoreWrite.
With the bug adding to store fails and the tests therefore do not run.
I think improving detection of SSH with combination of kPOSIXErrorBase may be long term fix. 
Alternatively, we can create new Keychain for the test so it is always writable independently of the the default one. 
For now, I verified that the tests run again when keychain is unlocked and the repro code runs on 10.15 & 11.0 


fixes #39603

wfurt committed:
hold ref to temp keychain on OSX to avoild premature cleanup

vcsjones commented on src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs:
```diff
@@ -220,25 +220,44 @@ internal static SafeKeychainHandle CreateOrOpenKeychain(string keychainPath, boo
                 Path.GetTempPath(),
                 Guid.NewGuid().ToString("N") + ".keychain");
 
             // Use a distinct GUID so that if a keychain is abandoned it isn't recoverable.
```
This comment is out of date since we aren't really using a GUID anymore.

vcsjones commented on src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs:
```diff
@@ -220,25 +220,44 @@ internal static SafeKeychainHandle CreateOrOpenKeychain(string keychainPath, boo
+                    out keychain);
+            }
 
+            utf8Passphrase.Fill(0);
```
```suggestion
            utf8Passphrase.Clear();
```

vcsjones commented on src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs:
```diff
@@ -220,25 +220,44 @@ internal static SafeKeychainHandle CreateOrOpenKeychain(string keychainPath, boo
 
-            byte[] utf8Passphrase = System.Text.Encoding.UTF8.GetBytes(tmpKeychainPassphrase);
+            // clear the binary bits.
+            random.Fill(0);
```
```suggestion
            random.Clear();
```

vcsjones commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/AppleCertificatePal.cs:
```diff
@@ -109,7 +111,20 @@ internal sealed partial class AppleCertificatePal : ICertificatePal
+                        keychain.DangerousAddRef(ref success);
+                        if (success)
+                        {
+                            ((AppleCertificatePal)ret)._tempKeychain = keychain;
```
I think you can change `ImportPkcs12` and `ImportPkcs12NonExportable` in `AppleCertificatePal.Pkcs12.cs` to return `AppleCertificatePal` so that the cast is not required.

vcsjones commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/AppleCertificatePal.cs:
```diff
@@ -109,7 +111,20 @@ internal sealed partial class AppleCertificatePal : ICertificatePal
                 using (keychain)
                 {
-                    return ImportPkcs12(rawData, password, exportable, keychain);
+                    var  ret = ImportPkcs12(rawData, password, exportable, keychain);
```
```suggestion
                    ICertificatePal ret = ImportPkcs12(rawData, password, exportable, keychain);
```

Or alternatively `AppleCertificatePal` if you decide to change the return type.

wfurt replied:
I don't know if the cast matters but I update it to return AppleCertificatePal.

bartonjs commented on src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs:
```diff
@@ -220,25 +220,44 @@ internal static SafeKeychainHandle CreateOrOpenKeychain(string keychainPath, boo
 
-            byte[] utf8Passphrase = System.Text.Encoding.UTF8.GetBytes(tmpKeychainPassphrase);
+            // clear the binary bits.
+            random.Fill(0);
```
```suggestion
            CryptographicOperations.ZeroMemory(random);
```

wfurt replied:
updated. Is that better for some reason then the Clear() @vcsjones was suggesting?

bartonjs replied:
Currently? No.  C compilers have always reserved the right (much to data security people's chagrin) to ignore memory writes that they can prove are never read again; so then calls like ZeroMemory get born to say "this one is important to clear".

If C#/.NET ever starts making those same optimizations, ZeroMemory will be immune.  We don't really expect to do that, but it's there just in case.  (It's implemented as a non-inlinable call to Clear :smile:)

wfurt replied:
that make sense. thanks for the explanation @bartonjs

bartonjs commented on src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs:
```diff
@@ -220,25 +220,44 @@ internal static SafeKeychainHandle CreateOrOpenKeychain(string keychainPath, boo
+                // Instead of true hexadecimal, we simply take lower and upper 4 bits and we offset them from ASCII 'A'
+                // to get printable form. We dont use managed string to avoid lingering copies.
+                utf8Passphrase[i*2] = (byte)((random[i] & 0x0F) + 65);
+                utf8Passphrase[i*2 + 1 ] = (byte)((random[i] >> 4) & 0x0F + 65);
```
```suggestion
                utf8Passphrase[i*2 + 1] = (byte)((random[i] >> 4) & 0x0F + 65);
```

bartonjs commented on src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs:
```diff
@@ -220,25 +220,44 @@ internal static SafeKeychainHandle CreateOrOpenKeychain(string keychainPath, boo
+            RandomNumberGenerator.Fill(random);
+
+            // Create hex-like UTF8 string.
+            Span<byte> utf8Passphrase =  stackalloc byte[randomSize * 2];
```
Let's put a null terminator on this.

```suggestion
            Span<byte> utf8Passphrase =  stackalloc byte[randomSize * 2 + 1];
            utf8Passphrase[randomSize * 2] = 0;
```

wfurt replied:
yah. that make sense even if we supply lenght. fixed.

wfurt committed:
feedback from review

vcsjones commented:
Did you confirm that the temp keychain files are getting cleaned up when the cert is disposed, or is that tested somewhere?

wfurt commented:
yes, I did. I modified you repro and added ReadLine after the import and added dispose followed by another ReadLine.
And the file is gone after dispose. I did not check any more complicated case when somebody would for example get and ref certificate handle or do some other tricks.

GrabYourPitchforks commented on src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs:
```diff
@@ -220,25 +220,46 @@ internal static SafeKeychainHandle CreateOrOpenKeychain(string keychainPath, boo
+
+            for (int i = 0; i < random.Length; i++)
+            {
+                // Instead of true hexadecimal, we simply take lower and upper 4 bits and we offset them from ASCII 'A'
```
Clever. :)

GrabYourPitchforks reviewed (commented):
Will let somebody more familiar with Apple Crypto sign off. But the modified logic in `CreateTemporaryKeychain` looks good to me!

wfurt commented:
/backport to release/5.0-rc2
